### PR TITLE
runtime: improvement/test struct for api mode watcher

### LINKED ISSUE #1528: Add better unit tests for `watcher`
### What problem are you trying to solve?

Avoiding releasing accidentally breaking watcher changes in the future.

0.92.0 had a change that passed 3 different types of tests; https://github.com/kube-rs/kube/issues/1524#issuecomment-2179054636 and still failed.

### Describe the solution you'd like

Create a `#[cfg(test)]` struct in `watcher.rs` and provide an injectable test implementation of `ApiMode` for it:
https://github.com/kube-rs/kube/blob/01f9c0c533db83b47aeb41bef783f151d4856dda/kube-runtime/src/watcher.rs#L158-L170

This would allow us to use this test struct to inject synthetic events (via e.g. `TestStruct::new(list, stream)`) that model actual apiserver responses (but without actually calling watch). This should allow us to verify a bunch of things from unit tests, such as (but not limited to):

- watcher does indeed call `list` N times (depending on page sizes), and then `watch`
- watcher does not call `list` in streaming lists
- watcher actually presents the union of the initial list pages and the next watch events (literally what 0.92.1 was made for https://github.com/kube-rs/kube/pull/1525 )
- watcher maintains chosen selectors for both list and watch calls
- watcher maintains correct flags on desyncs

It also lets us verify that that we are handling edge cases (currently we have very little error testing here).

### Describe alternatives you've considered

Extend mock tests in https://github.com/kube-rs/kube/blob/main/kube/src/mock_tests.rs#L29-L159

This has already been __somewhat__ done to avoid the previous case triggering, but would like something a little more robust ideally. There's little distinction between `list` and `watch` in these mock tests atm and I have not been able to test the watch side with it. If this test harness can be made more robust then this is also helpful, but ideally we should have unit tests in the are where the code is present (rather than rely on top level stuff too much).

### Target crate for feature

kube-runtime

### PR DESCRIPTION
**Pull request by daemonfire300**
## Motivation

Motivation is https://github.com/kube-rs/kube/issues/1528
Instead of reviving the stale PR I attempted a new approach from scratch.

## Solution

~~**IMPORTANT NOTE:** Contains unrelated files that are currently tracked in git because they have to be otherwise my devShell does not work (flake, nix, direnv). So feel free to ignore the flake.nix+flake.lock, if you want me to and like the nix devShell I can put that on a dedicated second PR. Also added a helper and git hooks for myself to have a Cargo.lock on hand, because I prefer having one when developing within a nix devShell, makes it more reproducible. I.e., I am going to remove those files from the PR later.~~

_It's not much and the change is NOT complete,_ before continuing I wanted to discuss the "interface" and behaviour of test struct(s). For example does the Duration i.e., ability to sleep/wait between calls to watch make sense, are we missing something to model a certain test. (Also I am not 100% sure whether the location of where I put the tests is the desired location for using that test struct.)

Happy to receive feedback.

### CODE REVIEW

**Comment by daemonfire300** [comment]
cc @mateiidavid since you have been on the old PR.

**Comment by daemonfire300** [comment]
I ll try to do a proper fix of the clippy issues tomorrow. Caught me by surprise because afaik `just clippy` does not run for `--test` and I get a bit of unrelated fallout from files I did not touch as well.

**Comment by doxxx93** [comment]
Don't worry about the `msrv` failure, it's unrelated to your PR. enum-ordinalize 4.4.0 (via [educe](https://docs.rs/educe/latest/educe/)) just raised its [MSRV to 1.89](https://github.com/magiclen/enum-ordinalize/commit/9185e0265ac715725b5a7b724523d6bfb4458867), which breaks our 1.88 check on any lockfile-less resolve.

cc) @clux should we consider [CARGO_RESOLVER_INCOMPATIBLE_RUST_VERSIONS](https://doc.rust-lang.org/cargo/reference/config.html#resolverincompatible-rust-versions) ?

**Comment by daemonfire300** [comment]
Any input on this so far?

**Comment by daemonfire300** [comment]
switched to draft as it technically is a DRAFT.

File: kube-runtime/src/watcher.rs [hunk 1]
```diff
@@ -475,6 +477,150 @@ where
...
+    }
+}
+
+#[cfg(test)]
+fn empty_list<K>() -> ObjectList<K>
+where
+    K: Clone,
+{
+    ObjectList {
+        types: kube_client::api::TypeMeta::default(), // TODO(juf): This is probably bad
```
Comment (clux) [inline kube-runtime/src/watcher.rs]: you should be able to use populate real values for TypeMeta if you constrain to `K: Resource`

File: kube-runtime/src/watcher.rs [hunk 2]
```diff
@@ -475,6 +477,150 @@ where
...
+        metadata: kube_client::api::ListMeta::default(), // TODO(juf): This is probably also bad
```
Comment (clux) [inline kube-runtime/src/watcher.rs]: this is probably ok 😄

File: kube-runtime/src/watcher.rs [hunk 3]
```diff
@@ -1004,6 +1150,115 @@ impl Backoff for DefaultBackoff {
...
+                resource_version: resource_version.map(str::to_string),
+                continue_: continue_token.map(str::to_string),
+                ..ListMeta::default()
+            },
+            items,
+        }
+    }
+
+    fn event_name(event: Result<Event<ConfigMap>>) -> String {
+        match event.unwrap() {
```
Comment (clux) [inline kube-runtime/src/watcher.rs]: i know this is tests, but it's a bit counter intuitive to have a fn that looks like a formatter call `unwrap`, unwrap should probably be lifted to the place calling `step`, and at that point, this could maybe have a generally useful `impl<K: Resource> Display for Event<K>`

File: kube-runtime/src/watcher.rs [hunk 4]
```diff
@@ -1004,6 +1150,115 @@ impl Backoff for DefaultBackoff {
...
+        }
+    }
+
+    #[tokio::test]
+    async fn listwatch_paginates_initial_list_before_starting_watch() {
+        let api = TestMode::new(
+            vec![
+                Ok(list(vec![cm("a", "1")], None, Some("first"))),
+                Ok(list(vec![cm("b", "2")], Some("2"), Some("second"))),
+                Ok(list(vec![cm("c", "3")], Some("3"), None)),
```
Comment (clux) [inline kube-runtime/src/watcher.rs]: i think this internal `list()` interface here is a little hard to read what the test is doing. partly this could be down to using realistic constants for rv and continue token, but maybe this needs named args

File: kube-runtime/src/watcher.rs [hunk 5]
```diff
@@ -475,6 +477,150 @@ where
     }
 }
 
+#[cfg(test)]
+enum Recording {
+    List(ListParams),
+    Watch(WatchParams, String),
+}
```
Comment (clux) [inline kube-runtime/src/watcher.rs]: I like the recorder and the way you are storing them within the test mode.

Some documentation on how it works and and what the strings represents would be useful.

File: kube-runtime/src/watcher.rs [hunk 6]
```diff
@@ -475,6 +477,150 @@ where
...
+
+#[cfg(test)]
+/// `TestMode` is the test-only "mock" implementation for [`ApiMode`].
```
Comment (clux) [inline kube-runtime/src/watcher.rs]: This whole set of logic is probably worth sticking into its own file, and then having the whole module gated by a `#[cfg(test)]`.

File: kube-runtime/src/watcher.rs [hunk 7]
```diff
@@ -1004,6 +1150,115 @@ impl Backoff for DefaultBackoff {
...
+
+    #[tokio::test]
+    async fn listwatch_paginates_initial_list_before_starting_watch() {
+        let api = TestMode::new(
+            vec![
+                Ok(list(vec![cm("a", "1")], None, Some("first"))),
+                Ok(list(vec![cm("b", "2")], Some("2"), Some("second"))),
+                Ok(list(vec![cm("c", "3")], Some("3"), None)),
+            ],
+            vec![Sequence::List(vec![Ok(WatchEvent::Added(cm("d", "4")))])],
```
Comment (clux) [inline kube-runtime/src/watcher.rs]: what was confusing me here for ages here was how the two vectors of inputs, and how one is for `list` and the other is for `watch`.

to test the worst watcher scenarios (particularly the older non-streaming lists cases), we can call `list`, then `watch`, and possibly `list` again on desyncs, so having interplay between two input vectors feels awkward.

File: kube-runtime/src/watcher.rs [hunk 8]
```diff
@@ -475,6 +477,150 @@ where
...
+                }
+            },
+            None => std::task::Poll::Ready(None),
+        }
+    }
+}
+
+#[cfg(test)]
+#[allow(clippy::unused_async_trait_impl)]
+impl<K> ApiMode for TestMode<K>
```
Comment (clux) [inline kube-runtime/src/watcher.rs]: This impl setup is great. Nice way to bookkeep.

File: kube-runtime/src/watcher.rs [hunk 9]
```diff
@@ -475,6 +477,150 @@ where
...
+    /// certain duration.
+    Wait(std::time::Duration),
+}
+
+#[cfg(test)]
+/// Implements [`futures::stream::BoxStream`] via [`futures::Stream`] for internal use via [`TestMode::watch`]
+pub struct TestStream<K> {
+    seq: RefCell<Vec<Sequence<K>>>,
+    waiting: Option<Pin<Box<tokio::time::Sleep>>>,
+}
```
Comment (clux) [inline kube-runtime/src/watcher.rs]: afaikt `TestStream` is exclusively dedicated for streaming things for `.watch`?

What happens in the case of errors and we need to return to an initial list?

File: kube-runtime/src/watcher.rs [hunk 10]
```diff
@@ -475,6 +477,150 @@ where
...
+///
+/// [`TestMode::watch_sequence`] is the fixed list of [`Sequence`]s `TestMode` returns. The
+/// behaviour, is similar to [`TestMode::fixture`] but it allows for simulating "waiting" periods,
+/// empty intermediary result and for returning a [`futures::stream::BoxStream`] implemented via [`TestStream`].
+struct TestMode<K>
+where
+    K: Clone + Debug + DeserializeOwned + Send + 'static,
+{
+    fixture: RefCell<Vec<kube_client::Result<ObjectList<K>>>>,
+    watch_sequence: RefCell<Vec<Sequence<K>>>,
```
Comment (clux) [inline kube-runtime/src/watcher.rs]: at minimal; these 2 fields could do with less confusing names.

but maybe it's alternatively worth considering combining these two vectors into a single vector that has information about how it's multiplexed internally?

**Review by clux — COMMENTED** [review]
thanks a lot for trying this. some initial comments.

this is definitely on the complex side, but the watcher is also complex, so it's not unexpected. the streaming setup and bookeeping seems very sensible, my biggest source of confusion and hesitation is with the input structures (for fixtures and sequence), and curious if there are alternatives along those lines.

i've put down some comments. mostly thinking out loud for now. still need some time to think about this. don't feel like you have to reply to all of it.

**Comment by daemonfire300** [comment]
> thanks a lot for trying this. some initial comments.
> 
> this is definitely on the complex side, but the watcher is also complex, so it's not unexpected. the streaming setup and bookeeping seems very sensible, my biggest source of confusion and hesitation is with the input structures (for fixtures and sequence), and curious if there are alternatives along those lines.
> 
> i've put down some comments. mostly thinking out loud for now. still need some time to think about this. don't feel like you have to reply to all of it.

Thank you very(!) much for the thorough review. Yes the tests are the weakest part. Will go over your points and (a) try to address them and (b) make things easier to understand.

File: kube-runtime/src/watcher.rs (hunk 10 above)
Comment (daemonfire300) [reply kube-runtime/src/watcher.rs]: Initially I kept them separate, because watch vs list return very different values:


```rust
async fn list(&self, lp: &ListParams) -> kube_client::Result<ObjectList<Self::Value>> {}
// vs
async fn watch(&self, wp: &WatchParams, version: &str) -> kube_client::Result<BoxStream<'static, kube_client::Result<WatchEvent<Self::Value>>>> {}
```

I agree with the naming, but unifying them into one I am less clear on the "why" angle. :+1:

File: kube-runtime/src/watcher.rs (hunk 9 above)
Comment (daemonfire300) [reply kube-runtime/src/watcher.rs]: Thanks for the remark: I think TestStream is insufficient. I re-read my implementation mutliple times now and TestStream only allows to model the outer happy path, not the outer Err case and internally it allows for both.

I.e., any call to `watch` will succeed if we would use it as is, which might not be desired/missing functionality. Will put that on my list of things to fix.

The part of "return an initial list" is, however, not clear to me. Either it is an error  _or_ we return a list, this should be this section https://github.com/daemonfire300/kube/blob/817092633873d7d84d41c3a8c10e64fd02acf07f/kube-runtime/src/watcher.rs#L674-L687 unless I am missing something

```mermaid
flowchart TD
    testMode["TestMode"]
    watch["watch"]
    sequences["inner Vec Sequence<...>"]
    error["Err(...)"]
    success["Ok(TestStream)"]

    testMode --- watch
    watch --> sequences
    watch -->|missing| error
    sequences -->|pop Vec Seq and put into Stream| success

    style error stroke:#ff3333,color:#ff3333,stroke-width:2px
    style success stroke:#20a83a,color:#20a83a,stroke-width:2px
```

Comment (daemonfire300) [reply kube-runtime/src/watcher.rs]: The behaviour of cosntructing the TestStream on a call to watch is broken. I have  Vec<Seq<..>> and pop from it, and then I think mistakenly create a single element Vec<Seq<..>> off of  that and put it into the stream which is (a) confusing and (b) not what I intended.

Comment (daemonfire300) [reply kube-runtime/src/watcher.rs]: I also need to fix the behaviour of `poll_next`, Poll::Ready(None) terminates the stream instead  of returning an empty result (the case for `Sequence::Empty`), which is confusing and "wrong".

File: kube-runtime/src/stub_watcher.rs [hunk 11]
```diff
@@ -0,0 +1,199 @@
...
+    async fn watch(
+        &self,
+        wp: &WatchParams,
+        version: &str,
+    ) -> kube_client::Result<futures::stream::BoxStream<'static, kube_client::Result<WatchEvent<Self::Value>>>>
+    {
+        self.recorder
+            .borrow_mut()
+            .push(Recording::Watch(wp.clone(), version.into()));
+        let seq = self.watch_sequence.borrow_mut().pop().into_iter().collect();
```
Comment (daemonfire300) [inline kube-runtime/src/stub_watcher.rs]: **### NOTE:** This still needs to addressed, because it's not necessarily proper behavior.

**Comment by clux** [comment]
I think this looks a lot better now. The tests read better, and are mostly understandable from reading the tests alone without having to dig deep into the implementation. You've addressed a lot of my concerns.

I don't have many more bikesheds on the actual code bits themselves, but I do think it would be interesting to try to add 2 more test cases to see if the abstraction can handle reality. Particularly around error handling and initial lists;

1. does the test setup support a streaming list init?
2. can we inject an error in the watch stream, and have it recover and get the next element?

File: kube-runtime/src/stub_watcher.rs [hunk 12]
```diff
@@ -0,0 +1,216 @@
...
+use crate::watcher::ApiMode;
+
+pub enum Recording {
+    List(ListParams),
+    Watch(WatchParams, String),
+}
+
+/// `TestMode` is the test-only "mock" implementation for [`ApiMode`].
+///
+/// [`TestMode::fixture`] is the fixed list of values `TestMode` returns, removing
```
Comment (clux) [inline kube-runtime/src/stub_watcher.rs]: nit; `fixture` is now only the arg in the constructor - should refer to the field in the struct.
Edit; same below.

File: kube-runtime/src/stub_watcher.rs [hunk 13]
```diff
@@ -0,0 +1,216 @@
...
+/// one element per call and returning it once none are left.
+/// This enables us to simulate different list scenarios.
+///
+/// [`TestMode::watch_sequence`] is the fixed list of [`Sequence`]s `TestMode` returns. The
+/// behaviour, is similar to [`TestMode::fixture`] but it allows for simulating "waiting" periods,
+/// empty intermediary result and for returning a [`futures::stream::BoxStream`] implemented via [`TestStream`].
```
Comment (clux) [inline kube-runtime/src/stub_watcher.rs]: nit; move to field in the struct for better doc colocation

File: kube-runtime/src/stub_watcher.rs [hunk 14]
```diff
@@ -0,0 +1,216 @@
...
+        self.recorder.borrow()
+    }
+}
+
+impl<K> TestMode<K>
+where
+    K: Clone + Debug + DeserializeOwned + Send,
+{
+    /// Arguments are mut because we reverse the order internally because we pop off the end
+    /// which means the first element to be returned would the last which would be unexpected.
```
Comment (clux) [inline kube-runtime/src/stub_watcher.rs]: if that's the reason you could maybe use a [VecDeque](https://doc.rust-lang.org/std/collections/struct.VecDeque.html) slightly more ergonomically. given perf isn't really a concern here.

File: kube-runtime/src/stub_watcher.rs [hunk 15]
```diff
@@ -0,0 +1,216 @@
...
+        self: std::pin::Pin<&mut Self>,
+        cx: &mut std::task::Context<'_>,
+    ) -> std::task::Poll<Option<Self::Item>> {
+        let this = self.get_mut();
+        if let Some(inner_fut) = this.waiting.as_mut() {
+            if !matches!(inner_fut.poll_unpin(cx), std::task::Poll::Ready(())) {
+                return std::task::Poll::Pending;
+            }
+            this.waiting = None;
+        }
```
Comment (clux) [inline kube-runtime/src/stub_watcher.rs]: how `waiting` works isn't really explained anywhere afaikt. maybe some comments about it would be good.

**Review by clux — COMMENTED** [review]
tiny nits. mostly about documentation.

File: kube-runtime/src/stub_watcher.rs (hunk 14 above)
Comment (daemonfire300) [reply kube-runtime/src/stub_watcher.rs]: It  was `VecDeque` initially, then I thought technically we only need to remove (`pop`) elements and to not need the whole "baggage" `VecDeque` comes with. I have little experience w.r.t. to `VecDeque` vs `Vec` with small data but a lot of objects. I guess I am still paranoid because I got stung very much by `HashMap` in a private project. But you suggestion sounds sensible. :+1:

File: kube-runtime/src/stub_watcher.rs [hunk 16]
```diff
@@ -0,0 +1,230 @@
...
+    }
+}
+
+impl<K> Default for Sequence<K> {
+    fn default() -> Self {
+        Sequence {
+            inner: VecDeque::new(),
+        }
+    }
+}
```
Comment (clux) [inline kube-runtime/src/stub_watcher.rs:150]: i think this `Default` can be derived (possibly with `educe` which we have in scope - since it's generic). doesn't really matter though.

File: kube-runtime/src/stub_watcher.rs (hunk 14 above)
Comment (clux) [reply kube-runtime/src/stub_watcher.rs]: end result looks a little cleaner :+1:

Comment (clux) [reply kube-runtime/src/stub_watcher.rs]: doc comment outdated now though (no mut, we don't reverse anymore)

File: kube-runtime/src/watcher.rs (hunk 9 above)
Comment (clux) [reply kube-runtime/src/watcher.rs]: maybe it helps to write some failing tests for the last cases for what you want first so you can hammer on the complex (sounding) behavior.

if the internals in the stream impl ends up being a bit ugly to accommodate the various test scenarios, that would still be a good tradeoff.

**Comment by daemonfire300** [comment]
> I think this looks a lot better now. The tests read better, and are mostly understandable from reading the tests alone without having to dig deep into the implementation. You've addressed a lot of my concerns.
> 
> I don't have many more bikesheds on the actual code bits themselves, but I do think it would be interesting to try to add 2 more test cases to see if the abstraction can handle reality. Particularly around error handling and initial lists;
> 
>     1. does the test setup support a streaming list init?
> 
>     2. can we inject an error in the watch stream, and have it recover and get the next element?

Built a reference test, and I think this is  broken by default by design, when I "re-use" my list case but port it to watch instead.
`step` infinity loops and the test hangs. TestMode is "depleted", but `step` continues so call it. Since I wrap TestMode using https://docs.rs/futures/latest/futures/stream/trait.StreamExt.html#method.fuse I am going to investigate this a bit more tomorrow. The primitive solution would probably to potential not use `.fuse()` to wrap the  stream and have it's own state management and  return some sort of error _OR_ call `step_trampolined` manually instead, but I am not sure whether this would be very pleasant ergonomics. In either case, I think this is by design as I said earlier, because a k8s controller using watch is not expecting the stream to "ever end" regularly which conflicts with the semantics of https://docs.rs/futures/latest/futures/stream/trait.Stream.html#tymethod.poll_next 

> Poll::Ready(None) means that the stream has terminated, and poll_next should not be invoked again.

I must say, this is exciting :+1: (to be read unironically)

**Comment by daemonfire300** [comment]
There was a bug in TestMode watch, fixed it, I was returning only the first element of a SequenceStep::List and discarding the rest.

As for the infinite loop on stream end: I am going to choose an error and return that so that any test using this does not hang. The alternative is having a specific error/code path which is guarded by #[cfg(test)] but I would not like to put test specific code mixed into the heart of the `watcher.rs`.

**Comment by daemonfire300** [comment]
most likely some helper/marker/tombstone error like this

```rust
fn exhausted_watch_sequence() -> kube_client::Error {
    kube_client::Error::ReadEvents(std::io::Error::new(
        std::io::ErrorKind::UnexpectedEof,
        "TestMode watch sequence exhausted",
    ))
}

```

File: kube-runtime/src/watcher.rs [hunk 17]
```diff
@@ -1003,7 +1029,230 @@ impl Backoff for DefaultBackoff {
...
+        }
+
+        let (event, _) = tokio::time::timeout(Duration::from_millis(100), step(&api, &config, state))
+            .await
+            .expect("step should return when the TestMode sequence is exhausted");
+        assert!(matches!(
+            event,
+            Err(Error::WatchFailed(kube_client::Error::ReadEvents(err)))
+                if err.kind() == std::io::ErrorKind::UnexpectedEof
+        ));
```
Comment (clux) [inline kube-runtime/src/watcher.rs:1155]: this error matching to handle the stream is "over" could perhaps be encasulated a bit better so the tests are more readable (since it's a magic value in the error).

File: kube-runtime/src/watcher.rs [hunk 18]
```diff
@@ -1003,7 +1029,230 @@ impl Backoff for DefaultBackoff {
...
+        tokio::time::advance(Duration::from_millis(1)).await;
+
+        assert!(matches!(
+            stream.next().await,
+            Some(Ok(WatchEvent::Added(config_map))) if config_map.metadata.name.as_deref() == Some("a")
+        ));
+    }
+
+    #[tokio::test]
+    async fn streaming_list_init() {
```
Comment (clux) [inline kube-runtime/src/watcher.rs:1086]: I think we need one final test here that verifies the working -> error -> resync setup. ie.

```
#[tokio::test]
async fn streaming_list_resync() {
```

where (pseudocode)

```
        for expected in [
            "InitApply(a)",
            "InitApply(b)",
            "InitDone",
            "Apply(a)",
            "Error(WatchFailed)",
            "InitApply(a),
        ] 
```

so we can simulate the internal step machinery resuming properly in the case of errors.

Comment (daemonfire300) [reply kube-runtime/src/watcher.rs:1086]: I made a  few more changes towards this, and I think the behaviour as of right now does not implement/handle what some of the doc-strings say. I.e., in case of an `Error` or an `WatchEvent::Error(...)`, `step_trampoline` does not reset the state and keeps the stream as well. Therefore on any subsequent call the same stream will be accessed, and no new call to `watch(...)` on the provided `Api/ApiMode` is made. This  would be the place where the last known/successfully received `resourceVersion` is present. The Underlying `http` client implementation only ever receives it.
There is one exception to what I said, and a partial exception at most: When we simulate getting a `WatchEvent::Error(err)` with `err.code == 410` we actually break the loop and (drop?) not re-use the existing stream because the new state is now `State::Empty`. However, _this_ is not the same as using the most recent resourceVersion for recovery afterwards, this just `resets`. This behaviour is mostly or completely ok, because when we do `InitialListStrategy::StreamingList`, we anyway get the previous events from the api-server.

But any error other than a `WatchEvent::Error(err) when err.code == 410` I am not sure behaves the way we intend it to behave for `watch()` (streaming lists).

I can elaborate a bit more and give a few more code-pointers if it helps understand the issue.

That said, I might be wrong / looking at this in the wrong angle.

cc @doxxx93 additional pair of eyes appreciated, w.r.t. to this.

Comment (daemonfire300) [reply kube-runtime/src/watcher.rs:1086]: Changes have not been pushed yet, because they do not really matter for what I described, i.e., they do not fix this.

Comment (doxxx93) [reply kube-runtime/src/watcher.rs:1233]: Your reading is right, but the resume isn't in the error branch by design. It takes two polls:

```
Watching, Some(Err(e))     -> emit WatchFailed, state stays Watching
Watching, None (next poll) -> InitListed { rv }
InitListed                 -> watch(Resumed, rv)
```

Both ways a watch dies give you that `None` right away. `request_events` is a `FramedRead`, and tokio-util returns `None` on the poll after an error; and for a `WatchEvent::Error` the apiserver closes the stream right after sending it (etcd3 `sendError` does `defer wc.cancel()`, so the handler's channel closes and it returns). So non-410 errors do resume from the last resourceVersion.

What misled you is the doc-string on `State::Watching`: it credits the error branch with what the end-of-stream branch actually does. I'll fix that wording separately so this PR stays about the harness.

The test is blocked by the stub, not the watcher: `TestStream::poll_next` yields `Some(Err(exhausted_watch_sequence()))` forever and never `Poll::Ready(None)`, so the watcher never leaves `State::Watching` and never calls `watch()` a second time. Let a sequence end in end-of-stream and the recorder will show the second `watch(..., rv)`.

One note on the expected sequence: the resume uses `WatchPhase::Resumed`, so `send_initial_events` is false (#1882) and the event after `Error(WatchFailed)` is `Apply(...)`, not `InitApply(a)`. `InitApply` only comes back through the 410 path, which resets to `State::Empty`.

Comment (doxxx93) [reply kube-runtime/src/watcher.rs:1233]: @clux one wording thing that fell out of this: the `State::Watching` doc-string reads as if the error branch does the restart.

Current:

```rust
/// If the connection is disrupted then we propagate the error but try to restart the watch stream by
/// returning to the `InitListed` state.
```

The error branch keeps the stream and stays put; it is the None on the following poll that goes to InitListed. That is what sent @daemonfire300 digging. Something like:

```
/// If the connection is disrupted we propagate the error and stay here. The stream ends on the
/// next poll (or hits the idle timeout), and that `None` is what returns us to `InitListed` to
/// resume from the last seen resourceVersion.
```

Wording ok with you?

Comment (daemonfire300) [reply kube-runtime/src/watcher.rs:1233]: @doxxx93 Thank you for checking.

My original implementation had `Poll::Ready(None)` see https://github.com/kube-rs/kube/pull/2022#issuecomment-5227549353 and exactly that caused issues.

Could you point me to exactly where I should return `Poll::Ready(None)` / when. Maybe I had a subtle bug in my first implementation, but it definitely caused the step/step_trampoline to call it in an infinite loop, iirc, but it is quite some time ago, I can check it again.

Comment (daemonfire300) [reply kube-runtime/src/watcher.rs:1233]: Ok I think I see your point, https://github.com/kube-rs/kube/pull/2022/changes#diff-9c6f7544f175ed853dd7281351f0c69fd2bdfcc439a5b74e520f2a2a25f6b9bfR205 this is what  you mean should be None.

Will check.

Comment (daemonfire300) [reply kube-runtime/src/watcher.rs:1233]: Pushed the changes. Now the only question is whether the tests look meaningful or are meaningful enough. I am sure there is potential to cover more cases (later?).

File: kube-runtime/src/watcher.rs [hunk 19]
```diff
@@ -1003,7 +1038,395 @@ impl Backoff for DefaultBackoff {
...
+                            vec![
+                                Ok(WatchEvent::Modified(config_map("a", LAST_VALID_RESOURCE_VERSION))),
+                                Ok(WatchEvent::Error(recoverable_err)),
+                            ]
+                            .into(),
+                        ),
+                        // After an error we expect the watcher to request its last(latest) known
+                        // resource version to be queried again.
+                        SequenceStep::List(
+                            vec![Ok(WatchEvent::Modified(config_map(
```
Comment (clux) [inline kube-runtime/src/watcher.rs]: looks great 👍

File: kube-runtime/src/watcher.rs [hunk 20]
```diff
@@ -1003,7 +1038,395 @@ impl Backoff for DefaultBackoff {
...
+                            )))]
+                            .into(),
+                        ),
+                    ]
+                    .into(),
+                ),
+                // After an error we expect the watcher to request its last(latest) known
+                // resource version to be queried again.
+                Sequence::new(
+                    vec![SequenceStep::List(
```
Comment (clux) [inline kube-runtime/src/watcher.rs]: i see every place you are doing `Sequence::new` you are passing a `something_vec.into()`. maybe we can move the `.into` into the `::new` constructor to make this a little less dense?

Comment (clux) [reply kube-runtime/src/watcher.rs]: The same seems true for the usage of `TestMode::new`

**Review by clux — APPROVED** [review]
This looks great to me now. Tests cover everything i wanted to cover, and we can always add more things later.

Added two tiny nits that I think could maybe be improved, but otherwise happy on my end.

**Review by clux — APPROVED** [review]
Once again, thanks a lot for this! I don't have any more even minor things to say so let's send this through and take future improvements incrementally.
